### PR TITLE
Frontier exploration integration with task allocator

### PR DESCRIPTION
**Pull request by rachelzh9**
- Working frontier_ta.py with greedy allocator

### CODE REVIEW

File: src/frontier_exploration/include/RRT.h [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        return min_index;
+    }
+
+    float dijkstra(int src, int dest)
+    {
+        using iPair = std::pair<float, int>;
+        std::priority_queue<iPair, std::vector<iPair>, std::greater<iPair>> pq;
+        std::vector<float> dist(nodes_.size(), std::numeric_limits<float>::max());
```
Comment (indraneelpatil) [inline src/frontier_exploration/include/RRT.h:95]: Does this dikstra's function work? Or can we remove it? Was this causing a seg fault?

Comment (indraneelpatil) [reply src/frontier_exploration/include/RRT.h:95]: Lets remove it if we are not using it

File: src/frontier_exploration/include/frontier_rrt_search.h [hunk 2]
```diff
@@ -11,13 +11,15 @@
...
 #include "mtrand.h"
+#include "RRT.h"
 
 #include "nav_msgs/OccupancyGrid.h"
 #include "geometry_msgs/PointStamped.h"
 #include "std_msgs/Header.h"
 #include "nav_msgs/MapMetaData.h"
 #include "geometry_msgs/Point.h"
 #include "visualization_msgs/Marker.h"
+#include "robosar_messages/rrt_path_cost.h"
```
Comment (indraneelpatil) [inline src/frontier_exploration/include/frontier_rrt_search.h:22]: Can we remove the service if we are not using it

File: src/frontier_exploration/include/frontier_rrt_search.h [hunk 3]
```diff
@@ -32,29 +34,33 @@ class FrontierRRTSearch
...
-    char ObstacleFree(std::pair<float, float> x_nearest, std::pair<float, float> &x_new);
+    char ObstacleFree(std::pair<float, float> &x_nearest, std::pair<float, float> &x_new);
 
-    int gridValue(std::pair<float, float> x);
+    int gridValue(std::pair<float, float> &x);
     std::pair<float, float> pixelsToMap(int x_pixel, int y_pixel);
 
+    void pruneRRT();
+
+    bool getPathCost(robosar_messages::rrt_path_cost::Request &req, robosar_messages::rrt_path_cost::Response &resp);
```
Comment (indraneelpatil) [inline src/frontier_exploration/include/frontier_rrt_search.h:49]: Does this service work?

Comment (indraneelpatil) [reply src/frontier_exploration/include/frontier_rrt_search.h:49]: If not using lets remove it

File: src/frontier_exploration/scripts/frontier_filter.py [hunk 4]
```diff
@@ -90,6 +95,49 @@ def init_markers(self):
 
         return points_clust
 
+    def is_valid_frontier(self, node):
+        if node[0] > -0.5 and node[0] < 12.0 and node[1] > -10.0 and node[1] < 2.0:
```
Comment (indraneelpatil) [inline src/frontier_exploration/scripts/frontier_filter.py:99]: What are these magic numbers? I am guessing this is a geofence? Please convert to macros or variables and add comments

File: src/frontier_exploration/scripts/frontier_filter.py [hunk 5]
```diff
@@ -90,6 +95,49 @@ def init_markers(self):
...
+            return True
+        return False
+
+    def obstacle_free(self, xnear, xnew):
```
Comment (indraneelpatil) [inline src/frontier_exploration/scripts/frontier_filter.py:103]: Can we rename this function? Obstacle free doesnt reflect what this function is doing at all, can we rename it to something like collision_check_connection

File: src/frontier_exploration/scripts/frontier_filter.py [hunk 6]
```diff
@@ -90,6 +95,49 @@ def init_markers(self):
...
+        if unk == 1:
+            out = -1  # unknown
+        if obs == 1:
+            out = 0  # occupied
+        if obs != 1 and unk != 1:
+            out = 1  # free
+
+        return out
+
+    def check_obs_intersection(self, centroid, label, cluster_labels, frontiers):
```
Comment (indraneelpatil) [inline src/frontier_exploration/scripts/frontier_filter.py:131]: Can we rename this function to something more apt too?

File: src/frontier_exploration/scripts/functions.py [hunk 7]
```diff
@@ -100,6 +100,7 @@ def discount(mapData, assigned_pt, centroids, infoGain, r):
 
 
 def pathCost(path):
+    r_region = 5
```
Comment (indraneelpatil) [inline src/frontier_exploration/scripts/functions.py:103]: What is this magic number? Add comments/macro

File: src/frontier_exploration/src/frontier_rrt_search.cpp [hunk 8]
```diff
@@ -11,31 +11,49 @@ FrontierRRTSearch::FrontierRRTSearch(ros::NodeHandle &nh) : nh_(nh)
     ns = ros::this_node::getName();
     ros::param::param<float>(ns + "/eta", eta, 10);
     ros::param::param<std::string>(ns + "/map_topic", map_topic, "/map");
+    ROS_INFO("map topic: %s", map_topic.c_str());
     ros::param::param<std::string>(ns + "/robot_leader", robot_leader, "agent1");
 
     map_sub = nh_.subscribe(map_topic, 100, &FrontierRRTSearch::mapCallBack, this);
     targets_pub = nh_.advertise<geometry_msgs::PointStamped>(ns + "/detected_points", 10);
     marker_pub = nh_.advertise<visualization_msgs::Marker>(ns + "_shapes", 10);
+    rrt_path_service_ = nh_.advertiseService("rrt_path_cost", &FrontierRRTSearch::getPathCost, this);
```
Comment (indraneelpatil) [inline src/frontier_exploration/src/frontier_rrt_search.cpp:20]: Are we using this service

File: src/frontier_exploration/src/frontier_rrt_search.cpp [hunk 9]
```diff
@@ -11,31 +11,49 @@ FrontierRRTSearch::FrontierRRTSearch(ros::NodeHandle &nh) : nh_(nh)
...
+}
+
 void FrontierRRTSearch::getRobotLeaderPosition()
 {
     ROS_INFO("Waiting for robot transform.");
     tf2_ros::Buffer tf_buffer_;
     tf2_ros::TransformListener tf_listener_(tf_buffer_);
     geometry_msgs::TransformStamped tf;
-    ros::Duration(5.0).sleep();
+    ros::Duration(1.0).sleep();
```
Comment (indraneelpatil) [inline src/frontier_exploration/src/frontier_rrt_search.cpp:51]: Can you add comment on why we need this sleep here

File: src/frontier_exploration/src/frontier_rrt_search.cpp [hunk 10]
```diff
@@ -11,31 +11,49 @@ FrontierRRTSearch::FrontierRRTSearch(ros::NodeHandle &nh) : nh_(nh)
...
     try
     {
-        ros::Time now = ros::Time(0);
+        ros::Time now = ros::Time::now();
         ROS_INFO("%s", robot_leader.c_str());
+        ros::Duration(3.0).sleep();
```
Comment (indraneelpatil) [inline src/frontier_exploration/src/frontier_rrt_search.cpp:56]: Why did you add this sleep here? No hardcoding :(

**Comment by indraneelpatil** [comment]
@rachelzh9  PR is too big, please make separate PRs in the future, dont keep adding things to same branch, make sure someone keeps reviewing your PR

File: src/frontier_exploration/include/RRT.h (hunk 1 above)
Comment (rachelzh9) [reply src/frontier_exploration/include/RRT.h:95]: we r keeping it just in case

File: src/frontier_exploration/include/frontier_rrt_search.h (hunk 3 above)
Comment (rachelzh9) [reply src/frontier_exploration/include/frontier_rrt_search.h:49]: no

File: src/frontier_exploration/scripts/frontier_filter.py (hunk 4 above)
Comment (rachelzh9) [reply src/frontier_exploration/scripts/frontier_filter.py:99]: geofence

**Comment by rachelzh9** [comment]
approve pls

**Review by indraneelpatil — COMMENTED** [review]
LGTM!
